unified-scheduler-pool/src/sleepless_testing.rs

```diff
@@ -0,0 +1,352 @@
+use std::{
+    io,
+    thread::{JoinHandle, Scope, ScopedJoinHandle},
+};
+
+#[allow(dead_code)]
+pub(crate) trait ScopeTracked<'scope>: Sized {
+    fn spawn_tracked<T: Send + 'scope>(
+        &'scope self,
+        f: impl FnOnce() -> T + Send + 'scope,
+    ) -> ScopedJoinHandle<'scope, T>;
+}
+
+pub(crate) trait BuilderTracked: Sized {
+    fn spawn_tracked<T: Send + 'static>(
+        self,
+        f: impl FnOnce() -> T + Send + 'static,
+    ) -> io::Result<JoinHandle<T>>;
+
+    #[allow(dead_code)]
+    fn spawn_scoped_tracked<'scope, 'env, T: Send + 'scope>(
+        self,
+        scope: &'scope Scope<'scope, 'env>,
+        f: impl FnOnce() -> T + Send + 'scope,
+    ) -> io::Result<ScopedJoinHandle<'scope, T>>;
+}
+
+#[cfg(not(test))]
+pub(crate) use sleepless_testing_dummy::*;
+#[cfg(test)]
+pub(crate) use sleepless_testing_real::*;
+
+#[cfg(test)]
+mod sleepless_testing_real {
+    use {
+        lazy_static::lazy_static,
+        std::{
+            cmp::Ordering::{Equal, Greater, Less},
+            collections::{HashMap, HashSet},
+            fmt::Debug,
+            sync::{Arc, Condvar, Mutex},
+            thread::{current, JoinHandle, ThreadId},
+        },
+    };
+
+    #[derive(Debug)]
+    struct Progress {
+        _name: String,
+        check_points: Vec<String>,
+        current_check_point: Mutex<String>,
+        condvar: Condvar,
+    }
+
+    #[derive(Debug)]
+    struct JustCreated;
```

Comment:
from library perspective, it is not obvious the user cannot use "JustCreated" as one of their check-points

Reply:
+1